-1

Tested at: 562b6e930f123c4705f2f48abf711243b262bd9c

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-4decca/789/summary.html

I found follow errors while testing this PR

**Failed tests:  ClangBuild**

* **Clang**:

I found compilation error while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```


<pre>>> Compiling  /build/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_0_X_2019-06-05-1100/src/Geometry/HGCalCommonData/src/HGCalDDDConstants.cc 
>> Compiling  /build/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_0_X_2019-06-05-1100/src/Geometry/HGCalCommonData/src/FastTimeParameters.cc 
>> Compiling  /build/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_0_X_2019-06-05-1100/src/Geometry/HGCalCommonData/src/HGCalWaferType.cc 
>> Compiling  /build/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_0_X_2019-06-05-1100/src/Geometry/HGCalCommonData/src/FastTimeDDDConstants.cc 
In file included from /build/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_0_X_2019-06-05-1100/src/Geometry/HGCalCommonData/src/HGCalGeomTools.cc:1:
/build/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_0_X_2019-06-05-1100/src/Geometry/HGCalCommonData/interface/HGCalGeomTools.h:48:27: error: constexpr variable 'factor_' must be initialized by a constant expression
  static constexpr double factor_ = 1.0/std::sqrt(3.0);
                          ^         ~~~~~~~~~~~~~~~~~~
/build/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_0_X_2019-06-05-1100/src/Geometry/HGCalCommonData/interface/HGCalGeomTools.h:48:41: note: non-constexpr function 'sqrt' cannot be used in a constant expression
  static constexpr double factor_ = 1.0/std::sqrt(3.0);
                                        ^
</pre>

--- a/Geometry/HGCalCommonData/interface/HGCalGeomTools.h
+++ b/Geometry/HGCalCommonData/interface/HGCalGeomTools.h
@@ -7,7 +7,7 @@
 
 class HGCalGeomTools {
 public:
-  HGCalGeomTools() {}
+  HGCalGeomTools();
   ~HGCalGeomTools() {}
 
   enum WaferType {WaferFull = 0, WaferFive=1, WaferChoptwo=2, WaferChopTwom=3, 
@@ -34,7 +34,7 @@ public:
                        std::vector<double> const& slope);
   static double radius(
       double z, int layer0, int layerf, std::vector<double> const& zFront, std::vector<double> const& rFront);
-  static std::pair<double,double> shiftXY(int waferPosition, double waferSize);
+  std::pair<double,double> shiftXY(int waferPosition, double waferSize);
   static double slope(double z, std::vector<double> const& zFront, std::vector<double> const& slope);
   static std::pair<double, double> zradius(double z1,
                                            double z2,
@@ -44,8 +44,8 @@ public:
       double xpos, double ypos, double r, double R, double rMin, double rMax, bool oldBug = false);
 
 private:
-  static constexpr double tol_    = 0.0001;
-  static constexpr double factor_ = 1.0/std::sqrt(3.0);
+  static constexpr double tol_ = 0.0001;
+  double  factor_;
 };
 
 #endif

--- a/Geometry/HGCalCommonData/src/HGCalDDDConstants.cc
+++ b/Geometry/HGCalCommonData/src/HGCalDDDConstants.cc
@@ -18,9 +18,6 @@
 //#define EDM_ML_DEBUG
 using namespace geant_units::operators;
 
-static const int maxType = 2;
-static const int minType = 0;
-
 HGCalDDDConstants::HGCalDDDConstants(const HGCalParameters* hp, const std::string& name)
     : hgpar_(hp), sqrt3_(std::sqrt(3.0)) {
   mode_ = hgpar_->mode_;
@@ -216,9 +213,9 @@ double HGCalDDDConstants::cellThickness(int layer, int waferU, int waferV) const
   int type = waferType(layer, waferU, waferV);
   if (type > 0) {
     if ((mode_ == HGCalGeometryMode::Hexagon8) || (mode_ == HGCalGeometryMode::Hexagon8Full)) {
-      thick = 10000.0 * hgpar_->cellThickness_[type-1];  // cm to micron
+      thick = 10000.0 * hgpar_->cellThickness_[type];  // cm to micron
     } else if ((mode_ == HGCalGeometryMode::Hexagon) || (mode_ == HGCalGeometryMode::HexagonFull)) {
-      thick = 100.0 * type;  // type = 1,2,3 for 100,200,300 micron
+      thick = 100.0 * (type+1);  // type = 1,2,3 for 100,200,300 micron
     }
   }
   return thick;
@@ -1235,10 +1232,10 @@ int HGCalDDDConstants::waferType(int layer, int waferU, int waferV) const {
   if ((mode_ == HGCalGeometryMode::Hexagon8) || (mode_ == HGCalGeometryMode::Hexagon8Full)) {
     auto itr = hgpar_->typesInLayers_.find(HGCalWaferIndex::waferIndex(layer, waferU, waferV));
     if (itr != hgpar_->typesInLayers_.end()) 
-      type = (1 + hgpar_->waferTypeL_[itr->second]);
+      type = hgpar_->waferTypeL_[itr->second];
   } else if ((mode_ == HGCalGeometryMode::Hexagon) || (mode_ == HGCalGeometryMode::HexagonFull)) {
     if ((waferU >= 0) && (waferU < (int)(hgpar_->waferTypeL_.size())))
-      type = hgpar_->waferTypeL_[waferU];
+      type = (hgpar_->waferTypeL_[waferU]-1);
   }
   return type;
 }

--- a/Geometry/HGCalCommonData/src/HGCalDDDConstants.cc
+++ b/Geometry/HGCalCommonData/src/HGCalDDDConstants.cc
@@ -361,7 +361,7 @@ int HGCalDDDConstants::getLayer(double z, bool reco) const {
   if (((mode_ == HGCalGeometryMode::Hexagon) || (mode_ == HGCalGeometryMode::HexagonFull)) && reco) {
     int indx = layerIndex(lay, false);
     if (indx >= 0)
-      lay = hgpar_->layerGroup_[indx];
+      lay = hgpar_->layerGroupO_[indx];
   } else {
     lay += (hgpar_->firstLayer_ - 1);
   }

--- a/Geometry/HGCalCommonData/src/HGCalDDDConstants.cc
+++ b/Geometry/HGCalCommonData/src/HGCalDDDConstants.cc
@@ -211,7 +211,7 @@ bool HGCalDDDConstants::cellInLayer(int waferU, int waferV, int cellU, int cellV
 double HGCalDDDConstants::cellThickness(int layer, int waferU, int waferV) const {
   double thick(-1);
   int type = waferType(layer, waferU, waferV);
-  if (type > 0) {
+  if (type >= 0) {
     if ((mode_ == HGCalGeometryMode::Hexagon8) || (mode_ == HGCalGeometryMode::Hexagon8Full)) {
       thick = 10000.0 * hgpar_->cellThickness_[type];  // cm to micron
     } else if ((mode_ == HGCalGeometryMode::Hexagon) || (mode_ == HGCalGeometryMode::HexagonFull)) {

--- a/Geometry/HGCalCommonData/src/HGCalGeomTools.cc
+++ b/Geometry/HGCalCommonData/src/HGCalGeomTools.cc
@@ -5,6 +5,7 @@
 #include <string>
 
 //#define EDM_ML_DEBUG
+HGCalGeomTools::HGCalGeomTools() : factor_(1.0/std::sqrt(3.0)) {}
 
 void HGCalGeomTools::radius(double zf,
                             double zb,

--- a/Geometry/HGCalGeometry/test/HGCalTestRecHitTool.cc
+++ b/Geometry/HGCalGeometry/test/HGCalTestRecHitTool.cc
@@ -49,12 +49,16 @@ private:
   const CaloGeometry* geom_;
   const int mode_;
   int layerEE_, layerFH_, layerBH_;
+  int layerEE1000_, layerFH1000_, layerBH1000_;
 };
 
 HGCalTestRecHitTool::HGCalTestRecHitTool(const edm::ParameterSet& iC)
     : geomToken_{esConsumes<CaloGeometry, CaloGeometryRecord>(edm::ESInputTag{})},
       geom_(nullptr),
-      mode_(iC.getParameter<int>("Mode")) {}
+      mode_(iC.getParameter<int>("Mode")) {
+   layerEE_     = layerFH_     = layerBH_     = 0;
+   layerEE1000_ = layerFH1000_ = layerBH1000_ = 0;
+}
 
 HGCalTestRecHitTool::~HGCalTestRecHitTool() {}
 
@@ -66,11 +70,13 @@ void HGCalTestRecHitTool::analyze(const edm::Event&, const edm::EventSetup& iSet
                                 : static_cast<const HGCalGeometry*>(
                                       geom_->getSubdetectorGeometry(DetId::HGCalEE, ForwardSubdetector::ForwardEmpty)));
     layerEE_ = (geomEE->topology().dddConstants()).layers(true);
+    layerEE1000_ = (geomEE->topology().dddConstants()).getLayer(10000., true);
     auto geomFH = ((mode_ == 0) ? static_cast<const HGCalGeometry*>(
                                       geom_->getSubdetectorGeometry(DetId::Forward, ForwardSubdetector::HGCHEF))
                                 : static_cast<const HGCalGeometry*>(geom_->getSubdetectorGeometry(
                                       DetId::HGCalHSi, ForwardSubdetector::ForwardEmpty)));
     layerFH_ = (geomFH->topology().dddConstants()).layers(true);
+    layerFH1000_ = (geomFH->topology().dddConstants()).getLayer(10000., true);
     if (mode_ == 0) {
       auto geomBH =
           static_cast<const HcalGeometry*>(geom_->getSubdetectorGeometry(DetId::Hcal, HcalSubdetector::HcalEndcap));
@@ -79,8 +85,12 @@ void HGCalTestRecHitTool::analyze(const edm::Event&, const edm::EventSetup& iSet
       auto geomBH = static_cast<const HGCalGeometry*>(
           geom_->getSubdetectorGeometry(DetId::HGCalHSc, ForwardSubdetector::ForwardEmpty));
       layerBH_ = (geomBH->topology().dddConstants()).layers(true);
+      layerBH1000_ = (geomBH->topology().dddConstants()).getLayer(10000., true);
     }
-    edm::LogVerbatim("HGCalGeom") << "Layers " << layerEE_ << ":" << layerFH_ << ":" << layerBH_ << std::endl;
+    edm::LogVerbatim("HGCalGeom") 
+      << "Layers " << layerEE_ << ":" << layerFH_ << ":" << layerBH_ 
+      << "\nLayer # at 1000 cm " << layerEE1000_ << ":" << layerFH1000_ << ":"
+      << layerBH1000_;
     for (int layer = 1; layer <= layerEE_; ++layer)
       edm::LogVerbatim("HGCalGeom") << "EE Layer " << layer << " Wafers "
                                     << (geomEE->topology().dddConstants()).wafers(layer, 0) << ":"

--- a/Geometry/HGCalGeometry/test/python/testHGCalRecHitTool_cfg.py
+++ b/Geometry/HGCalGeometry/test/python/testHGCalRecHitTool_cfg.py
@@ -4,7 +4,8 @@ process = cms.Process("PROD")
 process.load("SimGeneral.HepPDTESSource.pdt_cfi")
 
 process.load("Configuration.Geometry.GeometryExtended2023D17Reco_cff")
-process.load("Configuration.Geometry.GeometryExtended2023D17_cff")
+#process.load("Configuration.Geometry.GeometryExtended2023D28Reco_cff")
+#process.load("Configuration.Geometry.GeometryExtended2023D41Reco_cff")
 process.load('FWCore.MessageService.MessageLogger_cfi')
 
 if hasattr(process,'MessageLogger'):
@@ -36,6 +37,7 @@ process.maxEvents = cms.untracked.PSet(
 
 process.prodTest = cms.EDAnalyzer("HGCalTestRecHitTool",
                                   Mode = cms.int32(0),
+#                                 Mode = cms.int32(1),
                                   )
 
 process.p1 = cms.Path(process.generator*process.prodTest)

